File: algorithms/289. Game of Life/solution.py
class Solution(object):
    def gameOfLife(self, board):
        """
        :type board: List[List[int]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        if not board:
            retrn
        r = len(board); c = len(board[0])
        for i in xrange(r):
            for j in xrange(c):
                l = 0
                if i-1 >= 0 and j-1 >= 0 and board[i-1][j-1] > 0:
                    l += 1
                if i-1 >= 0 and board[i-1][j] > 0:
                    l += 1
                if i-1 >= 0 and j+1 < c and board[i-1][j+1] > 0:
                    l += 1
                if j+1 < c and board[i][j+1] > 0:
                    l += 1
                if i+1 < r and j+1 < c and board[i+1][j+1] > 0:
                    l += 1
                if i+1 < r and board[i+1][j] > 0:
                    l += 1
                if i+1 < r and j-1 >= 0 and board[i+1][j-1] > 0:
                    l += 1
                if j-1 >= 0 and board[i][j-1] > 0:
                    l += 1
                if board[i][j] > 0:
                    board[i][j] = l or 1
                else:
                    board[i][j] = -l
        for i in xrange(r):
            for j in xrange(c):
                if 0 < board[i][j] < 2 or board[i][j] > 3:
                    board[i][j] = 0
                elif 2 <= board[i][j] <= 3 or board[i][j] == -3:
                    board[i][j] = 1
                else:
                    board[i][j] = 0

File: algorithms/289. Game of Life/solution.cc
class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        int m = board.size(), n = m ? board[0].size() : 0;
        for ( int i = 0; i < m; ++i ) {
            for ( int j = 0; j < n; ++j ) {
                int count = 0;
                for ( int x = i-1; x <= i+1; ++x ) {
                    for ( int y = j-1; y <= j+1; ++y ) {
                        if ( 0 <= x && x < m && 0 <= y && y < n ) {
                            count += board[x][y] & 1;
                        }
                    }
                }
                if ( count == 3 || count - board[i][j] == 3 ) {
                    board[i][j] |= 2;
                }
            }
        }
        for ( int i = 0; i < m; ++i ) {
            for ( int j = 0; j < n; ++j ) {
                board[i][j] >>= 1;
            }
        }
    }
};
